ublox_derive/tests/test.rs

```diff
@@ -1,138 +1,83 @@
-use std::convert::TryInto;
-use ublox_derive::{ubx_packet};
-use num_derive::{FromPrimitive, ToPrimitive};
-use num_traits::{FromPrimitive, ToPrimitive};
+use ublox_derive::{UbxDefineSubTypes, UbxPacketRecv, UbxPacketSend};
 
-#[ubx_packet]
-struct TestPacket {
-    field1: u32,
-    field2: u32,
-    field3: u8,
-    field4: u16,
-    field5: i32,
+trait UbxPacket {
+    const class: u8;
+    const id: u8;
+    const fixed_length: Option<u16>;
 }
 
-#[test]
-fn foo() {
-    assert_eq!(std::mem::size_of::<TestPacket>(), 15);
+#[derive(UbxPacketRecv, UbxDefineSubTypes)]
+#[ubx(class = 1, id = 2, fixed_len = 28)]
+#[repr(packed)]
+/// Geodetic Position Solution
+struct NavPosLLHRaw {
+    itow: u32,
+    #[ubx(map_type = f64, scale = 1e-7, alias = lon_degrees)]
+    lon: i32,
+    lat: i32,
+    #[ubx(map_type = f64, scale = 1e-3, alias = height_ellipsoid_meters)]
+    height_ellipsoid: i32,
+    height_msl: i32,
+    /// Horizontal Accuracy Estimate
+    horizontal_accuracy: u32,
+    vertical_accuracy: u32,
 }
 
-#[ubx_packet]
-struct TestPacket2 {
-    field1: u32,
-    field2: u8,
-    field3: u32,
+#[derive(UbxPacketRecv, UbxDefineSubTypes)]
+#[repr(packed)]
+#[ubx(class = 1, id = 3, fixed_len = 16)]
+struct NavStatusRaw {
+    itow: u32,
+    #[ubx(enum {
+	NoFix = 0,
+	DeadReckoningOnly = 1,
+	Fix2D = 2,
+	Fix3D = 3,
+	GpsDeadReckoningCombine = 4,
+	TimeOnlyFix = 5,
```

Comment:
What advantage to you get from specifying this inline versus referencing enums specified elsewhere?

Some enums are used in multiple places, for example this `gpsFix` field of NavStatus has the same values as `gpxFix` in NavSol, and ostensibly the same values as the `fixType` field in NavPvt (the latter seems to be more GNSS-oriented vs. the GPS-specific type).

(here and elsewhere, I'm looking at the u-blox 7 protocol specification, that's what I have handy)

Reply:
> What advantage to you get from specifying this inline versus referencing enums specified 
> elsewhere?

The whole idea to have "safe" interface. User will not work with bits directly only via restricted interface. To get this to work "ublox_derive" have to know all members of enum,
plus it can generate proper enum:
```rust
#[repr(TYPE_FROM_PACKET)]
enum Foo {
  // user specified values
  Unknown(TYPE_FROM_PACKET)
}
```
Plus without knowledge about enum variants it is impossible to generate
TYPE_FROM_PACKET -> Foo converter like:
```rust
let x: TYPE_FROM_PACKET
match x {
  0 => Foo::Val1,
}
```
it is possible to convert Foo -> TYPE_FROM_PACKET via `as TYPE_FROM_PACKET`, but not other way around.

I don't take into consideration "shared" enums,
for them I suppose it is possible:
```rust
#[derive(UbxDefineSubTypes)]
enum Foo {
}
```

to pass information about this enum into "ublox_derive".
But "inline" syntax is more nice way, because of we can not edit already existing types
via proc_macro compiler plugin. So user have to add things like `Unknown` variant by hands.

Reply:
Well, the "user" in this case is users of `ublox_derive`, which should really just be contributors to this project. End users shouldn't be needing to define their own enums. So, I'm not too hurt if we have to hand-specify our own `Unknown` field for enums where it makes sense.

Deriving a trait on the enum to make safety should work just fine, for example the `FromPrimitive` trait I believe gives us a safe way to do it (although instead of `Unknown`, it throws an error, which I think might be preferable - if the device gives us a packet we can't fully understand, should we parse it successfully?)

Reply:
I give another thought, I suppose it is almost impossible for proc_macro script.

```
#[derive(UbxDefineSubTypes)]
enum Foo {
}
#[derive(UbxDefineSubTypes)]
struct Pack {

} 
```

ublox_derive get the `Pack` and `Foo` via separate functions calls.
And the order also unspecified,
```rust
#[proc_macro_derive(UbxDefineSubTypes, attributes(ubx))]
pub fn ubx_packet_subtypes(input: TokenStream) -> TokenStream {
```

In other words it is possible to ubx_packet_subtypes can be invoked in parallel, in separate threads for `Foo` and `Pack`, and during processing `Pack` we can not get information about what is `Foo`.

But it would be possible via `build.rs`, the usage of build script I suppose allow us even do better, like combining checking of all defined packages automatically, so we can generate:
```
match (class, id) {
(NavStatus::class, NavStatus::id) => ...,
(NavPosLLH::class, NavPosLLH::id) => ...,
```
without human involvement.

> Deriving a trait on the enum to make safety should work just fine, for example the 
> FromPrimitive trait I believe gives us a safe way to do it

Actually I am not fond of small crates for 2+2. 
I have reasonably small project workspace of 40 private packages and dependencies tree from crates.io something like ~350 crates. And I have bunch of duplicates which I can not fix,
like three num_traits: 0.1.43, 0.2.11...

Why use third party create for something as simple as:
```
match n {
        1 => Some(MyEnum::V1),
        2 => Some(MyEnum::V2),
        _ => None,
    }
```

> if the device gives us a packet we can't fully understand, should we parse it successfully?

I suppose we should parse it successfully in case of  unknown bits are "reserved" one.
The purpose of "reserved" is exactly is this - changing of protocol without breaking software.
Why oppose it? I suppose the developer who care about protocol changing, also not expect this behavior.

If user cares about protocol version he/she can request version via UBX protocol and check.
I suppose reporting error for 10 types of packets that they contains bits in reserved area is not nice thing.

Reply:
> And I have bunch of duplicates which I can not fix

Coming from a C/C++/Python world, I can see how this is unfortunate. But how bad is it here in Rust? Unless you do `cargo clean` a lot, you would never even know.

> I suppose we should parse it successfully in case of unknown bits are "reserved" one

There's two types of "reserved" we have to care about:
- Reserved fields being populated. For example, the `reserved1` field of NavPvt, or bits 7 through 3 of the `valid` bitfield in NavPvt.
- Receiving new values for existing enums, e.g. if we got 0x13 for the `fixType` field of NavPvt.

In the first case, I agree, we should ignore set bits and pass them on. In the second case, though, I'm less convinced - if a user is relying on fixType, it complicates their code to have to handle the `Unknown` case. And what can they do when they get that? At best, they can error anyway. I don't want user code to look like this:
```
match fixType {
   NoFix => { fail() }
   DeadReckoning => { fail() }
   Unknown(0x13) => {
      // ublox doesn't know about this packet @TODO: Move upstream
      do_something()
   }
   Unknown(_) => {
      fail();
   }
}
```
Or whatever. If a user needs to get value 0x13 out of `fixType`, I want them to post an issue on this repo saying "I need to get 0x13 from this packet." (they can, of course, still do the modification locally so they don't have to wait for a fix) That way their code stays cleaner, and we get notified (and ideally a PR!) whenever ublox changes up the protocol.

> In other words it is possible to ubx_packet_subtypes can be invoked in parallel

As long as the enum and the struct are separated by a trait the enum implements (such as FromPrimitive), is this a problem?

Reply:
> But how bad is it here in Rust?

I suppose it is not related to specific language. 
Let's suppose you introduce third party packet/library/crate for your python/c++/rust program.
At now you responsible for watch that third party component and backport or update this component if some bad bug is found. Now back to Rust I have to watch 3!! version of crate X,
and if bad bug is found I need check if which one of is affected or may all of them.
In other words if have in dependency tree 3 version of the same crate the amount of work is similar like I have 3 different crates with the same idea behind but with different implementation.

> And what can they do when they get that? At best, they can error anyway. I don't want user code to look like this:

I suppose you choose bad example. If we talk about `NavStatus::fixType` then manual explicitly tell do not use this field to detect receiver state, so user just print it to log or show user via GUI/web page, and then use `NavStatus::flags` to decide is it possible to use related `NAV-POSLLH` messages or not. And I suppose it perfectly fine to show Unknown(..) in log or GUI if program have no idea what these bits mean and it doesn't affect the main purpose - safe way to receive spatial data.

> As long as the enum and the struct are separated by a trait the enum implements (such as FromPrimitive), is this a problem?

I suppose it can do the trick, but gathering all packets types into list for parsing still the problem.

Reply:
> Now back to Rust I have to watch 3 versions of crate X

Let's say you depend on crates A, B, and C, which each depend on a different version of X. In version 1.2.1, X is found to have a bug. From your point of view, what's the difference between X having this bug and each of A, B, and C having this bug? You don't care that there is a bug in X, you only care how it affects the operation of A, B, and C. If A, B, and C all do what you need them to, why would you want to upgrade X even if it has the bug?

But let's contrast this to the situation where there is no X, and A, B, and C each wrote their own code to do X. If a bug is found in how C does X, that bug may very well be in both A and B as well - after all, they are also doing X - except now we have no way of knowing whether A and B have this bug as well. It seems like it'd be better to centralize all of the code into X, so that everyone can have a single source of truth for how to do X correctly.

The problem is mitigated even further by using tools which auto-update dependencies. I gotta set something like that up.

> I suppose you choose bad example

haha, it seems I did choose a poor example. Perhaps `status` in NavDgps would be a better example? Or, `month` in NavPvt?

The latter example begging the question of - should this layer catch obvious validation errors, such as day > 31 or minute > 59?

Reply:
> Perhaps status in NavDgps

Yes, this looks better. We definitely need report error here.

> Or, month in NavPvt?
> The latter example begging the question of - should this layer catch obvious validation errors, such as day > 31 or minute > 59?

I suppose no, if don't return proper type.
For example let's we validate and return `u8` as month.
Paranoid user (like I am) have to revalidate it, because of `u8` is from 0..255,
and even if in the range 0..11 or 1..12 , it still may be invalid like February 30.
So parser will just use CPU cycles for nothing, because of user have to revalidate it.

So I suppose we return numbers without validation or we return something like `chrono::DateTime`. But this require some kind of "compaction" syntax, 
to describe how to combine all fields to one. Plus dependency on `chrono`,
because of obviously leap year calculation is error prone.

Reply:
> It seems like it'd be better to centralize all of the code into X

But there is centralization into X. In general only last version of X receive fix.
And if have A depend on X1, B depend on X2, and C depend on X3,
only C receive proper fix that I can get via `cargo update`,
I need manually patch B and A via `override` in cargo section.

So in ideal world it may be works, but in reality each crate in dependency tree is problem.
Because always there is probability that you have become it's maintainer.

Reply:
Yeah, for dates it gets complicated quickly. So I guess we can take it on a case-by-case basis, which is fine - some enums get `Unknown` fields (such as `fixType`), some don't (such as `NavDgps::status`).

Reply:
But at least you _can_ patch B and A, and you know that they have the bug, and you know how to fix the bug. If A, B, and C each have separate implementations of X, and C's implementation has a bug, how would you ever find the bug in A or B?

From the point of view of the owners of A, B, and C, reimplementing X may be a lot of work - not only in writing the code, but in finding all the roadblocks that X has already solved for them.

For complicated things, the benefit grows (such as HTTP servers) - for simple things, the risk is low (what sort of bug are you worried about in `num_traits` that will affect your end-user code?)

Reply:
> But at least you can patch B and A, and you know that they have the bug

I don't know that they have the bug. I only know that X3 has bug and it was fixed.
I have to dig X1 and X2 to find it out. In that case if A and B provide similar functionality to X and don't use X as dependency 
After each major release may be completely rewritten code. So I don't see big difference,
if A and B "embed" functionality of X I can in similar way look at them.

> For complicated things, the benefit grows 

Don't get me wrong, I am against tiny crates like `num_traits` that doesn't contain any complex functionality, for such kind of crates there are more minuses than pluses.

> what sort of bug are you worried about in num_traits that will affect your end-user code?

It provides some function, I am not sure that all corner cases for all possible number types are covered. And it is "trait" package. That mean that:

```rust
//A crate
impl num_traits::XYZ for A {}
//B crate
impl num_traits::XYZ for B {}
 //my crate
fn some_func<T: num_traits::XYZ>(_: T) {}
```
doesn't work, because of num_traits::XYZ for different versions are not the same (of course if "semver trick" was not used).

Reply:
> I can also reread [A and B's] code

This does not sound easier than checking to see if they depend on some particular version of a crate. Checking for old crate versions can even be done automatically, without having to look at their code.

> doesn't work, because of num_traits::XYZ for different versions are not the same

Sure, and this is inconvenient, but if B had implemented their own `B::XYZ` trait, you'd be no better off.